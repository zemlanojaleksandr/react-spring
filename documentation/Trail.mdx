---
name: Trail
route: /trail
order: 5
---

import { PropsTable } from 'docz'
import { Trail, config, animated } from '../src/targets/web'
import { RewindSpringProvider, RewindSpring } from './resources/RewindSpring.js'
import CodeMirror from './resources/Codemirror'
import DemoGrid from '../examples/components/DemoGrid'
import Demo from '../examples/components/Demo'
import examples from '../examples/components/examples'

# Trail

```jsx
import { Trail } from 'react-spring'
```

<RewindSpringProvider>
  <p>
    A <code>Trail</code> animates the first item of a list of elements, the rest form a natural trail and follow their previous sibling.
  </p>
  <p><span className="highlight">It takes a list of items of any type, and their keys</span>. The latter defaults to <code>item => item</code>, which, if your items are self-sufficient as a key is often good enough.
  </p>
  <p>
     <span className="highlight">You provide a single function-child</span> that receives an item and additionally its index. The full signature looks like:
  </p>

  <CodeMirror value={`(item, index) => props => ReactNode`}/>
    
  <p style={{ marginTop: '1em' }}>
    TLDR, items in, keys if necessary, present the item that comes out and apply the animated props you receive.
  </p>

  <div className="code-table">
    <CodeMirror value={`const items = [...]
  
<Trail
  items={items} keys={item => item.key}
  from={{ transform: 'translate3d(0,-40px,0)' }}
  to={{ transform: 'translate3d(0,0px,0)' }}>
  {item => props =>
    <span style={props}>{item.text}</span>
  }
</Trail>`}/>
    <RewindSpring>
      {x => (
        <>
          <animated.div style={{ opacity: x.interpolate({ range: [0.0,1.0], output: [0,1] }), transform: x.interpolate({ range: [0.0,1.0], output: [-40,0], extrapolate: 'clamp' }).interpolate(x => `translate3d(0,${x}px,0)`) }}>h</animated.div>
          <animated.div style={{ opacity: x.interpolate({ range: [0.2,1.0], output: [0,1] }), transform: x.interpolate({ range: [0.2,1.0], output: [-40,0], extrapolate: 'clamp' }).interpolate(x => `translate3d(0,${x}px,0)`) }}>e</animated.div>
          <animated.div style={{ opacity: x.interpolate({ range: [0.4,1.0], output: [0,1] }), transform: x.interpolate({ range: [0.4,1.0], output: [-40,0], extrapolate: 'clamp' }).interpolate(x => `translate3d(0,${x}px,0)`) }}>l</animated.div>
          <animated.div style={{ opacity: x.interpolate({ range: [0.6,1.0], output: [0,1] }), transform: x.interpolate({ range: [0.6,1.0], output: [-40,0], extrapolate: 'clamp' }).interpolate(x => `translate3d(0,${x}px,0)`) }}>l</animated.div>
          <animated.div style={{ opacity: x.interpolate({ range: [0.8,1.0], output: [0,1] }), transform: x.interpolate({ range: [0.8,1.0], output: [-40,0], extrapolate: 'clamp' }).interpolate(x => `translate3d(0,${x}px,0)`) }}>o</animated.div>
        </>
      )}
    </RewindSpring>
  </div>
</RewindSpringProvider>

## Props

<p>
  <span className="highlight">Trail accepts <i>all</i> spring properties</span>: native, from, immediate, onRest, and so on.
</p>

<PropsTable of={Trail} />

## Demos

<DemoGrid padding={0}>
  {examples.filter(data => data.tags.includes('trails')).map(data => (
    <Demo
      key={data.name}
      {...data}
      link={`https://github.com/drcmda/react-spring/tree/master/examples/demos/${data.name}`}
      import={import('../examples/demos/' + data.name)}
    />
  ))}
</DemoGrid>
