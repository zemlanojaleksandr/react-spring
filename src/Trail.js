import React from 'react'
import PropTypes from 'prop-types'
import Spring from './Spring'
import { config as springConfig } from './shared/constants'
import { toArray } from './shared/helpers'

export default class Trail extends React.PureComponent {
  static propTypes = {
    /** Item keys (the same keys you'd hand over to react in a list). If you specify items, keys can be an accessor function (item => item.key) */
    keys: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    ).isRequired,
    /** Base values, optional */
    from: PropTypes.object,
    /** Animates to ... */
    to: PropTypes.object,
    /** Item keys (the same keys you'd hand over to react in a list). If you specify items, keys can be an accessor function (item => item.key) */
    keys: PropTypes.oneOfType([PropTypes.func, PropTypes.array, PropTypes.any]),
    /** An array of items to be displayed, use this if you need access to the actual items when distributing values as functions (see above) */
    items: PropTypes.oneOfType([PropTypes.array, PropTypes.any]).isRequired,
    /** A single function-child that receives the individual item and return a functional component (item, index) => props => view) */
    children: PropTypes.func.isRequired,
    /** When true the trailing order is switched, it will then trail bottom to top */
    reverse: PropTypes.bool,
  }

  static defaultProps = { keys: item => item }

  first = true
  instances = new Set()
  hook = (instance, index, length, reverse) => {
    // Add instance to set
    this.instances.add(instance)
    // Return undefined on the first index and from then on the previous instance
    if (reverse ? index === length - 1 : index === 0) return undefined
    else return Array.from(this.instances)[reverse ? index + 1 : index - 1]
  }

  render() {
    const {
      items,
      children,
      from = {},
      initial,
      reverse,
      keys,
      delay,
      onRest,
      ...props
    } = this.props

    const array = toArray(items)
    return toArray(array).map((item, i) => (
      <Spring
        onRest={i === 0 ? onRest : null}
        key={typeof keys === 'function' ? keys(item) : toArray(keys)[i]}
        from={this.first && initial !== void 0 ? initial || {} : from}
        {...props}
        delay={(i === 0 && delay) || undefined}
        attach={instance => this.hook(instance, i, array.length, reverse)}
        children={props => {
          const child = children(item, i)
          return child ? child(props) : null
        }}
      />
    ))
  }

  componentDidUpdate(prevProps) {
    this.first = false
    if (prevProps.items !== this.props.items) this.instances.clear()
  }
}
