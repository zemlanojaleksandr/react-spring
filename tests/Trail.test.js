import React from 'react'
import { shallow } from 'enzyme'
import Trail from '../src/Trail'

const fakeChild = () => {
  ;<div>fake</div>
}

// These snapshots won't reveal much, but they at least will catch hard errors.
describe('Trail', () => {
  it('smoke/regression tests', () => {
    const wrapper = shallow(
      <Trail keys={[0, 1]}>
        <fakeChild />
        <fakeChild />
      </Trail>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
