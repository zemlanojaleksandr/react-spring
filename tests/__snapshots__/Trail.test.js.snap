// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trail smoke/regression tests 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Trail
    keys={
      Array [
        0,
        1,
      ]
    }
  >
    <fakeChild />
  </Trail>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
